### PR TITLE
Enable themes and consolidate colors, gradients

### LINKED ISSUE #4085: Merge hydrogen palettes (dt-, tm-), rename
## Summary

As a result of #4049 we will have two coexsiting colour palettes while we rebrand. This issue is to merge the old `dt-` prefixed colours with the new `tm-` and rename them.

### PR DESCRIPTION
**Pull request by substrae**
🤖 Resolves #4085 #5620.

## 👋 Introduction

This PR adds support for Hydrogen's theme capabilities, meaning that we can now control configured colors, gradients, shadows, etc. using theme values on the `html data-h2` attribute. In practice, this means that ALL of our UI is now themeable, so that all components are accessible in the DT or IAP branding.

Beyond that, Hydrogen's theme functionality also enables us to use theme-specific modifiers in Hydrogen attributes: `base:iap()` to target individual themes. For example, this means you can use `display` to hide and show elements based on a particular theme.

As a result, configured colors and gradients have been consolidated as well!

## Theme provider

@esizer has built out a preliminary hook that allows us to specify the appropriate theme based on the app's current `context`.

- The backend admin UI defaults to the `admin` theme
- The applicant facing profile and application UI defaults to Hydrogen's core settings
- The IAP homepage defaults to the `iap` theme

Any pages that are added to the existing contexts should receive the correct theme automatically, though can control these themes manually if the need ever arises (for example, applying the `iap` theme to the application pages).

### Colors

Because the theme functionality automatically applies styles for you, there is no longer any reason to have theme-specific prefixes. As a result, they have been consolidated in the Hydrogen configuration file and in the code to the following options:

```
primary    | success | black
secondary  | warning | gray
tertiary   | error   | white
quaternary |         |
quinary    |         |
```

image

image

No more `dt-primary` or `purple`. Generally speaking you'll mostly be using the `primary` and `secondary` colors alongside the stoplight and shade options.

### Gradients

For similar reasons, the gradients have also been consolidated:

- `main-linear`
- `home-footer-linear`
- `primary-transparent-linear`

## Enhanced components

Some components have been visually updated alongside the theme updates for better consistency. These are purely visual changes and don't impact functionality.

- `missingSkills`
- `pagination`
- `dialog` and `alertDialog`
- `alert` and `toast`

## Enhanced component stories

Storybook has also had a bit of an upgrade. @esizer has added functionality that allows you to swap between themes at the click of a button to ensure that your components look great and are accessible in all situations.

image

When creating a new component story, you now have the ability to specify a `default` theme for the component if it doesn't necessarily make sense to have it default to the base theme. This is specifically helpful for IAP specific components and pages.

Alongside this enhanced functionality, some component stories have been enhanced to reflect the updated colors. A few examples can be found below.

<details>
  <summary>Call-to-action</summary>
  
</details>

<details>
  <summary>Alert</summary>
  
</details>

<details>
  <summary>Button</summary>
  
</details>

<details>
  <summary>Card</summary>
  
</details>

<details>
  <summary>Pill</summary>
  
</details>

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

Chromatic link: https://61e099a1bb1465003a73d3ce-ennmuuagda.chromatic.com

1. Rerun the setup script to ensure you have the latest and greatest packages.
2. Run `npm run dev` to compile the latest styles
3. Check the admin site for the correct theme.
4. Check the applicant site for the correct theme.
5. Check the IAP homepage for the correct theme.
6. Make sure nothing looks amiss.
7. Run `npm run storybook` and check out the new theme toggle and how it modifies components.
8. Ensure components and other stories are a-okay.

### CODE REVIEW

**Comment by mnigh** [comment]
Is there an issue that this PR resolves that includes acceptance criteria?

**Comment by substrae** [comment]
@esizer I think of the changes we made broke the Jest tests:

image

**Comment by mnigh** [comment]
Is the "secondary" **Create a new talent request** meant to look like a link now?

https://github.com/GCTC-NTGC/gc-digital-talent/blob/4067ab19a2bf031b31368f6b5b030746da9c677c/apps/web/src/pages/SearchRequests/RequestConfirmationPage/RequestConfirmationPage.tsx#L146-L157

## feature/themes branch
### [chromatic link](https://www.chromatic.com/component?appId=61e099a1bb1465003a73d3ce&csfId=pages-request-confirmation-page&buildNumber=4406&k=63fcddcc737c40c074d7f6b5-1200-interactive-true&h=9&b=-3)

Screen Shot 2023-02-27 at 16 36 15


## main branch
### [chromatic link](https://www.chromatic.com/component?appId=61e099a1bb1465003a73d3ce&csfId=pages-request-confirmation-page&buildNumber=4404&k=63fcddcc737c40c074d7f6b5-1200-interactive-true&h=10&b=-4)

Screen Shot 2023-02-27 at 16 36 25

**Comment by petertgiles** [comment]
This looks great, Josh! :rocket:   I've started by just going through the Chromatic diffs.

I noticed a few spots where contrast and readability seem like they could be improved.  Could you please double check these to see if you agree?
image
image
image
image
image
image

The links seems a bit droopy:
image

Did you mean to remove this button?
image

The dark mode snapshots broke:
image
image
image

Should this button be right-aligned?
image

**Comment by BrookeCameron** [comment]
@petertgiles just a quick catch users should be able to select both Status AND Non-status First Nations - you are either one or the other

**Comment by petertgiles** [comment]
> users **should not** be able to select both Status AND Non-status First Nations - you are either one or the other

Yep, it's just Storybook where you can set any state you want.  In the app it's not possible.

**Comment by mnigh** [comment]
> > users **should not** be able to select both Status AND Non-status First Nations - you are either one or the other
> 
> Yep, it's just Storybook where you can set any state you want. In the app it's not possible.

There is an open issue for the storybook aspect of it though, see: https://github.com/GCTC-NTGC/gc-digital-talent/issues/5686

**Comment by BrookeCameron** [comment]
Thanks @petertgiles & @mnigh !

**Comment by substrae** [comment]
@mnigh It's not that `secondary` buttons have changed, it's that I made that specific button `inline` rather than `outline`

@petertgiles 
- In terms of contrast, I made the following changes, but otherwise things passed based on my contrast tests on WebAIM
  - Disabled buttons in general are iffy based on the reduction in opacity that we're applying - I don't remember being involved in this design specifically, but I've reduced the effect ever so slightly to help
  - I've reduced the darkness of wells and increased the value of the optional form label
  - Links on the profile weren't implemented as button components properly, so those have been updated and, as a result, are darker and bolder
  - I've darkened the equity text in the well
  - I've overhauled the application status element to use the otherwise universal `<Well>`, so the contrast there is significantly better (see below)
- I gave the application cards a bit of interim love (they're being updated based on the new dashboard anyway, but this addressed the droopy link (see below)
- Same idea here - the type was set to `link` instead of `button`, so that's been changed
- Dark mode snapshots breaking is deliberate and temporary until a second PR can be made to address dark mode specifically
- I think this button was always right aligned and we couldn't tell because the label didn't wrap to a new line, but the bold text made it just large enough; I centred it

Application status:

image

Application card: 

image

**Comment by petertgiles** [comment]
> * Dark mode snapshots breaking is deliberate and temporary until a second PR can be made to address dark mode specifically

Could you please open an issue for this and link it back here?

File: apps/web/src/components/UserProfile/ProfileSections/AboutSection.tsx [hunk 1]
```diff
@@ -219,28 +219,6 @@ const AboutSection: React.FC<AboutSectionProps> = ({
           ""
         )}
       </div>
-      {!firstName &&
```
Comment (petertgiles) [inline apps/web/src/components/UserProfile/ProfileSections/AboutSection.tsx:222]: Was this intentional?  It seems out of scope.

**Review by petertgiles — COMMENTED** [review]
I'm about half way through my code review. :smile:  Some comments so far...

File: apps/web/src/components/Hero/Hero.tsx [hunk 2]
```diff
@@ -48,6 +49,7 @@ const Hero = ({
   crumbs,
   children,
   centered = false,
+  contentAlignment = "center",
```
[the remark this answers is not in the document]
Comment (substrae) [reply apps/web/src/components/Hero/Hero.tsx:52]: Ah, I was also thinking about this, but this property is unique in the sense that it centres the text in the child card element, not the entire banner. The granular control was important in this specific case.

File: apps/web/src/components/UserProfile/ProfileSections/AboutSection.tsx (hunk 1 above)
Comment (substrae) [reply apps/web/src/components/UserProfile/ProfileSections/AboutSection.tsx:222]: This section had two null states, styled totally differently from each other, that were both showing at the same time. The redundancy seemed enough for me to remove it as a part of my clean up, but if it was doing something specific that I'm missing, we can absolutely restore it.

Comment (substrae) [reply apps/web/src/components/UserProfile/ProfileSections/AboutSection.tsx:222]: image

**Comment by esizer** [comment]
We seem to be having some issue with colour contrast related to primary and secondary in some places when they are used in combination with light colours.

## Cards

- **Current**: 2.21:1
- **Expected**: 3:1

Screenshot 2023-03-01 084152

card-header


## Links

- **Current**: 3.21:1
- **Expected**: 4.5:1

Screenshot 2023-03-01 084319

links

## Primary Text

- **Note**
- This specific example could be fixed by bumping the font size by 0.66px but I believe it is not always used with large text so may be better just to darken it a bit.

- **Current**: 4.34:1
- **Expected**: 4.5:1

Screenshot 2023-03-01 084537

primary-text

## Optional Label

- **Current**: 3.28:1
- **Expected**: 4.5:1

Screenshot 2023-03-01 084634
optional

## Profile Links

- **Current**: 2.61:1
- **Expected**: 4.5:1

Screenshot 2023-03-01 084739
profile-links

**Review by petertgiles — APPROVED** [review]
No more concerns from me.  Looks really nice!  :artist: 

I pushed this up to the dev server so lets give the other designers a chance to review before it gets merged.

**Comment by esizer** [comment]
> Optional label:

Hmm odd, when i test the optional label I get `rgb(139, 139, 139)` which is `#8B8B8B`, not `#5d5d5d`.

Screenshot 2023-03-01 094618

> Links and profile links:

For this, I got `rgb(0, 190, 178)` which is `#00BEB2`, not `#006B64`

Screenshot 2023-03-01 094537

---

I'm not sure what is going on here because these do not match the css variables, but those are the colours being rendered to the screen 🤔

**Review by esizer — APPROVED** [review]
Sweet! 🍭

**Comment by RM-1978** [comment]
Thanks for letting me do a quick A11y peer review of the Home page! 

Here are a few thoughts: 
 

> - The "create a profile" section within primary content of the main section of the page should have an H2 or H3 title called "Your profile" to mark it clearly and be consistent across the UI. 
>  - The function of the all photographic images may be considered a little bit more than decorative if we consider that we are trying to communicate that GC Digital talent includes a wide representation of diverse cultures, gender expressions, age, etc. If this is the case, then their image function (e.g. conveying the different types of cultural communities who can access GCDT) should have text alternatives. ( this might be a larger ticket later)
> - Although the contrast between the red background and the black text for the "Get ready by creating a profile" link is acceptable for the minimum contrast ratio - 2.1 AA (WCAG 1.4.3), a contrast ratio of 7+ (AAA) is preferred by neurodiverse, including some with low vision.
> - Although the text in 11pt in the learn More section passes  AA conformance criteria ( WCAG 1.4.3) for minimum contrast ratio, it might be worth bumping up the font size to 14 pt so that the text has a better contrast ratio of at least 7:1 for “regular” sized text and at least 4.5:1 for large text, unless the text is purely decorative.
> - The colour contrast ratio between the aqua marine text against the white background to indicate the language component's hover state does not seem to pass requirements for minimum AA contrast ratio conformance (WCAG 1.4.3).
> - The colour contrast ratio between the aqua marine text against the white background to indicate the hover state of the hyperlinks in the "Check it out" section does not seem to pass requirements for minimum AA contrast ratio conformance (WCAG 1.4.3).
> - The colour contrast ratio between the aqua marine text against the white background to indicate the hover state of the first 5 hyperlinks in the footer section does not seem to pass requirements for minimum AA contrast ratio conformance (WCAG 1.4.3).

**Comment by substrae** [comment]
@RM-1978 Are these comments specific to the homepage in the report summary? Or should I be looking elsewhere?

**Comment by RM-1978** [comment]
> @substrae I tested the home page targeted by the report summary (question of timelines). That said, if the observation for black against green or the convention for the link hover state repeats on other pages, it might be good to address any repeat instances we can. I can do a bit more testing with more pages on UAT at a later point in the work.

**Comment by substrae** [comment]
Okay, the link hover colors have been addressed, but I'm not sure which 11px text is being referred to, as all the text in that section on the homepage is `16px` minimum on smaller devices and `18px` on larger ones.

**Comment by esizer** [comment]
> Okay, the link hover colors have been addressed, but I'm not sure which `11px` text is being referred to, as all the text in that section on the homepage is `16px` minimum on smaller devices and `18px` on larger ones.

The "Date modified" in the footer is 10.8pt/14,4px

**Comment by substrae** [comment]
All of our `caption` size text is that font size, so unless we decide to remove it sitewide, I'm not sure what the best path forward is there.

**Comment by esizer** [comment]
> All of our `caption` size text is that font size, so unless we decide to remove it sitewide, I'm not sure what the best path forward is there.
> 
> Edit: that said, I would also argue font size is out of scope for this particular PR.

I agree, font size is def out of scope. Something we may need to address at a later point.

**Comment by substrae** [comment]
Sounds like a plan! Given that it meets AA in this case, and the design is intentionally reducing that element's visual priority, I think we're okay as is in this specific case, but we can absolutely review `caption` in a separate discussion if it's causing issues elsewhere.

**Comment by RM-1978** [comment]
Thanks @substrae and @esizer.  So, to get the information I needed to test the font-size, I worked with another graphic designer to get another perspective. I think that the font-size issue might have been due to the sample text extraction we did so we could test the contrast and the font-size on my off-network device before testing (gain time constraints). In the end, what this tells me is that I need the CCA installed on my network device. 

If we want to scope out font-size from this PR, then please note that I recommend considering this part of  WCAG 1.4.3 at some point in the near future. Perhaps in a next Accessiblity epic. Thanks!
